### DISCUSSION TITLE
A question about locks and file leases in FUSE

### DISCUSSION THREAD

**Discussion by ChrysoliteAzalea** [libfuse/libfuse#853] [General]
Hello everyone! I'd like to ask about fcntl functions in the Linux kernel designed to protect against some race conditions such as file locking and leasing. File locking and leasing allow sequential reading and writing operations to be scheduled in a way that they do not interfere with each other. A shared lock means that a piece of data is being read, so any other party is free to read or "sharedly-lock" the piece of data, but processes wanting to modify or exclusively lock the data should wait. An exclusive lock means that a piece of data is being modified, and any other party wanting to read, write or lock the data in any way should wait.

The Linux kernel implements advisory locking (means that processes can wait on a lock if they want to, but are not preventing from working with locked data ignoring the lock) and mandatory locking (enforced by the kernel, attempts to work with locked data would either block or return the ```EAGAIN``` error). The mandatory lock is considered [unreliable](https://lwn.net/Articles/667215/) (because if mandatory locking overlaps with a begin of a ```write``` system call, the lock may succeed but the modification may "cut in line", for example). Because of this, a better way of protecting the data access known as [file leasing](https://manpages.ubuntu.com/manpages/jammy/man2/fcntl.2.html) is introduced. Leasing is also mandatory (enforced by the kernel), but unlike mandatory locking, it cannot be applied while incompatible file descriptors exist (writable FDs in case of read lease or *any* other FDs in case of write lease), block the concurrent open rather the read/write, and have a time limit (once a lease break is initiated, the lease holder is sent ```SIGIO``` and the lease can be automatically removed or downgraded by the kernel).

I wonder, does FUSE provide any way to handle leasing and locking? For example, it can be very useful for network file systems and shared cloud storages to avoid concurrent data modifications (however, in that case it has to be enforced by the server).

**Comment by bsbernd** [comment]
I don't think fuse has and you should not rely on it anyway, see below.  I think what you actually want to use are fuse notifications, although they need to get extended for network file system (on my companies TODO list).

https://man7.org/linux/man-pages/man2/fcntl.2.html

Mandatory locking
       Warning: the Linux implementation of mandatory locking is
       unreliable.  See BUGS below.  Because of these bugs, and the fact
       that the feature is believed to be little used, since Linux 4.5,
       mandatory locking has been made an optional feature, governed by
       a configuration option (CONFIG_MANDATORY_FILE_LOCKING).  This
       feature is no longer supported at all in Linux 5.15 and above.

BUGS section
Mandatory locking
       The Linux implementation of mandatory locking is subject to race
       conditions which render it unreliable: a [write(2)](https://man7.org/linux/man-pages/man2/write.2.html) call that
       overlaps with a lock may modify data after the mandatory lock is
       acquired; a [read(2)](https://man7.org/linux/man-pages/man2/read.2.html) call that overlaps with a lock may detect
       changes to data that were made only after a write lock was
       acquired.  Similar races exist between mandatory locks and
       [mmap(2)](https://man7.org/linux/man-pages/man2/mmap.2.html).  It is therefore inadvisable to rely on mandatory
       locking.
